### PR TITLE
feat(PTP-1013): ipfs feature, upload file and folder

### PR DESCRIPTION
**Pull request by salimtb**

### CODE REVIEW

File: src/lib/IPFS/ipfs.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    if (!projectSecret) {
+      throw new Error(
+        errorLogger({
+          location: ERROR_LOG.location.Ipfs_constructor,
+          message: ERROR_LOG.message.no_infura_projectSecret_supplied,
+        }),
+      );
+    }
+
+    const auth = 'Basic ' + Buffer.from(projectId + ':' + projectSecret).toString('base64');
```
Comment (kalote) [inline src/lib/IPFS/ipfs.js:42]: could be using the utils file / toBase6() function

File: src/lib/IPFS/ipfs.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    this.#ipfsClient = ipfsClient({
+      url: ipfsUrl,
+      headers: {
+        authorization: auth,
+      },
+    });
+  }
+
+  async uploadFile({ source }) {
```
Comment (kalote) [inline src/lib/IPFS/ipfs.js:52]: I don't think we need intermediary methods here. You can maybe directly call "save", "saveDirectory" from the exposed method, no?

File: src/lib/IPFS/ipfs.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    const uploadedDirectory = [];
+    for await (const file of this.#ipfsClient.addAll(globSource(source, '**/*'), {
+      wrapWithDirectory: true,
+    })) {
+      uploadedDirectory.push(file);
+    }
+
+    return [...uploadedDirectory].pop().cid.toString();
```
Comment (kalote) [inline src/lib/IPFS/ipfs.js:114]: what do you return here? Only the folder cid? shouldn't we return all the files' cid?

File: src/services/httpService.js [hunk 4]
```diff
@@ -7,18 +7,17 @@
 /* eslint-disable */
```
Comment (kalote) [inline src/services/httpService.js:8]: missing logger implementation in this file

File: test/ipfs.test.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    addAll: mockedAddAllCall,
+
+    pin: {
+      rm: mockedRemoveCall,
+    },
+  })),
+}));
+
+loadEnv();
```
Comment (kalote) [inline test/ipfs.test.js:39]: move around line 7

File: usage.js [hunk 6]
```diff
@@ -34,6 +34,16 @@ const acc = new Auth({
 
 const sdk = new SDK(acc);
 
+const ipfs = new IPFS({
```
Comment (kalote) [inline usage.js:37]: this doesn't work: where does IPFS is defined?

You're missing the export in index.js i will say :)

Comment (kalote) [reply usage.js:37]: BTW, we should integrate the IPFS with Auth class, to have an "all-in-one" definition of the credentials parameters.

Comment (kalote) [reply usage.js:37]:
```
const acc = new Auth({
  privateKey: process.env.WALLET_PRIVATE_KEY,
  projectId: process.env.INFURA_PROJECT_ID,
  secretId: process.env.INFURA_PROJECT_SECRET,
  rpcUrl: process.env.EVM_RPC_URL,
  chainId: 5,
  ipfsUrl: process.env.INFURA_IPFS_ENDPOINT,
});

const sdk = new SDK(acc);
const hash = sdk.ipfs.uploadFile({
  source: './integration-test/ipfs-test/infura.png',
});
```

WDYT?

**Review by kalote — CHANGES_REQUESTED** [review]
implementation of logger / export to make the ipfs function available through sdk

File: usage.js (hunk 6 above)
Comment (salimtb) [reply usage.js:37]: yes you're right, i have to export in index.js, for the integration of IPFS module in the same class, we're following the recommandation of adrien delaroche for now , that why it's a separated module

File: src/lib/IPFS/ipfs.js (hunk 3 above)
Comment (salimtb) [reply src/lib/IPFS/ipfs.js:114]: we are returning only the hash , cid is an instance of the class `CID`, and to get the hash you have to call the method toString() of this class

File: integration-test/ipfs.test-integration.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  });
+
+  it('should upload a directory', async () => {
+    const hash = await ipfs.uploadDirectory({
+      source: folder,
+    });
+
+    uploadedDirectoryHash = hash;
+
+    expect(hash).not.toBe(null);
```
Comment (MqllR) [inline integration-test/ipfs.test-integration.js:39]: The file CID will be always the same. I think we can also check the result of the uploadDirectory to ensure we have the expected array of hashes.

File: src/lib/IPFS/ipfs.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+ * MIT Licensed
+ */
+/* eslint-disable */
+import { ERROR_LOG, errorLogger } from '../error/handler.js';
+import fs from 'fs';
+import { create as ipfsClient, globSource } from 'ipfs-http-client';
+import { toBase64 } from '../utils.js';
+
+export default class IPFS {
+  #ipfsClient;
```
Comment (MqllR) [inline src/lib/IPFS/ipfs.js:13]: IMO this one should be public so everyone could use the ipfsClient on their own.

File: usage.js (hunk 6 above)
Comment (MqllR) [reply usage.js:37]: IPFS is not under web3 project so it's another pair of credentials

Comment (kalote) [reply usage.js:37]: even with another set of credentials, it's our kitchen to manage. We don't want the dev user to bother with increased complexity. We can do something like:
```
const acc = new Auth({
  privateKey: process.env.WALLET_PRIVATE_KEY,
  projectId: process.env.INFURA_PROJECT_ID,
  secretId: process.env.INFURA_PROJECT_SECRET,
  rpcUrl: process.env.EVM_RPC_URL,
  chainId: 5,
  ipfs: {
    url: process.env.INFURA_IPFS_ENDPOINT,
    projectId: process.env.INFURA_IPFS_PROJECT_ID,
    secretId: process.env.INFURA_IPFS_SECRET_ID,
  }
});
```
WDYT?

**Review by kalote — CHANGES_REQUESTED** [review]
dependencies

**Pull request closed without merging**